### ISSUE TITLE
Consolidate metrics under the sidekiq.* namespace

### ISSUE THREAD

**Issue by weppos** [sidekiq/sidekiq#6337]
Ruby version: -
Rails version: -
Sidekiq / Pro / Enterprise version(s): 7.x

[Sidekiq Pro sends metrics](https://github.com/sidekiq/sidekiq/wiki/Pro-Metrics) under the following namespaces:

```
sidekiq.*
jobs.*
batches.*
leader.election
```

This behavior is undesirable as, when blended into a large organization account, you can easily miss metrics. There are options to namespace the metrics, but the namespace is applied to all metrics.

So for instance, if I set the `"sidekiq"` namespace, as it follows:

```
config.dogstatsd = -> { Datadog::Statsd.new("1.2.3.4", "8125", namespace: "sidekiq") }
```

I will obtain the following:

```
sidekiq.sidekiq.*
sidekiq.jobs.*
sidekiq.batches.*
sidekiq.leader.election
```

Ideally, all metrics are reported under a consistent `"sidekiq"` prefix. Adjusting the namespace will pre-pend the prefix.

After some internal configuration, we obtained something similar applying a namespace, and remapping "sidekiq.sidekiq.*" block to "sidekiq.*".

<img width="726" alt="Screenshot 2024-06-20 at 10 30 33" src="https://github.com/sidekiq/sidekiq/assets/5387/9aaf54ed-4b37-4145-a411-8a7689c42b3e">

**Comment by javierjulio** [comment]
@weppos I'm dealing with the same situation setting up metrics and I was curious what you are using currently for the `namespace:` keyword argument in the `Datadog::Statsd` instance? Did you end up using `namespace: "sidekiq.my-app"` to get things working?

**Comment by javierjulio** [comment]
Worth noting that in the [Datadog documentation for updating the Agent configuration file](https://docs.datadoghq.com/integrations/sidekiq/#configuration) it does suggest matching against a "sidekiq.sidekiq." style prefix so that may have to be changed.

```
# dogstatsd_mapper_cache_size: 1000  # default to 1000
dogstatsd_mapper_profiles:
  - name: sidekiq
    prefix: "sidekiq."
    mappings:
      - match: 'sidekiq\.sidekiq\.(.*)'
        match_type: "regex"
        name: "sidekiq.$1"
      - match: 'sidekiq\.jobs\.(.*)\.perform'
        name: "sidekiq.jobs.perform"
        match_type: "regex"
        tags:
          worker: "$1"
      - match: 'sidekiq\.jobs\.(.*)\.(count|success|failure)'
        name: "sidekiq.jobs.worker.$2"
        match_type: "regex"
        tags:
          worker: "$1"
```

**Comment by weppos** [comment]
I used the recommended setup in the Agent. I set `namespace: "sidekiq"` in the `Datadog::Statsd` instance, then used the mapper to rewrite `sidekiq.sidekiq` entries (line 1 of 3).

The mappers 2 and 3 are technically unnecessary, if you use most recent versions of Sidekiq. They were used to map metrics where the worker name was part of the metric name, into tags. But the use of tag was already introduced in Sidekiq codebase one or two major versions ago.

**Comment by javierjulio** [comment]
@weppos thank you. That's most helpful. A final clarification, Did you happen to only have one Sidekiq running or multiple Sidekiq's for different apps? I'm guessing the former as to why the namespace could just be `namespace: "sidekiq"` vs it including the app/service name?

**Comment by weppos** [comment]
> @weppos thank you. That's most helpful. A final clarification, Did you happen to only have one Sidekiq running or multiple Sidekiq's for different apps? I'm guessing the former as to why the namespace could just be `namespace: "sidekiq"` vs it including the app/service name?

I anticipated your question here 😉  https://github.com/sidekiq/sidekiq/issues/6406#issuecomment-2318779397

TL;DR: we have only one app, but it is irrelevant. Even if we had more, we would not prefix metrics by app. Instead, we use tagging which is a better way to organize the information.

We have other services like Postgres for which we run multiple instances. Same deal, the metrics are not prefixed by the instance, instead we tag each source and we use the same metric name. This allows us to view data both aggregated or isolated per tag.

**Comment by weppos** [comment]
Note this setup is also the one somehow implicitly assumed by the default Dashboard that Datadog makes you available when you enable the Sidekiq integration. And this is how they structure Dashboard and metrics across most services.

**Comment by mperham** [comment]
Thanks @weppos, I forgot about tags. Tags offer different way of organizing your metrics and may lead to a different set of capabilities when viewing/querying/filtering. YMMV!

**Comment by mperham** [comment]
Not that I'm aware of.

**Comment by mperham** [comment]
I have not started work on it yet, I’m still finishing up 7.3.

**Comment by mperham** [comment]
@travisTheOrange start by opening your own issue. This is not your issue.

All tech support is via issues or email. Phones are terrible for support because I can’t see the code.

**Comment by weppos** [comment]
Do you mind to elaborate a bit what changes are being implemented in 8.0? Do you add `sidekiq.*` in front of every sidekiq name?

**Comment by mperham** [comment]
Right now I have removed any `sidekiq.` prefix from all metrics. The metrics now look like:

```
history.processed
history.queue.size
jobs.success
jobs.failure
batches.create
batches.success
```

What do you think? Should there be a common prefix for all Sidekiq-sourced metrics? Is that the purpose of the `:namespace` option or is this different?[^1] In other words, what should our metric naming conventions and best practices be?

[^1]: Customers can use `Sidekiq::Pro.metrics` to send their own app metrics too. If we use the namespace option, those app metrics will get that namespace too.

**Comment by weppos** [comment]
> Should there be a common prefix for all Sidekiq-sourced metrics?

I think so. Looking at my DD account, I see that each service has its own default prefix. So it seems reasonable to have all Sidekiq metrics being prefixed by `sidekiq.`.

The namespace is generally an additional prefix that is being customized. For instance, some organizations would use namespace to prefix the app name when running Sidekiq on multiple instances.

It's ultimately up to you, but from what I've seen with all the services we use, it is reasonable that on a standard setup the namespace is empty. But if you don't prefix them by sidekiq, then users will be forced to always use namespaces to avoid name clashing.

**Comment by mperham** [comment]
I'm not sure of the right thing.

Datadog does [document](https://docs.datadoghq.com/getting_started/tagging/unified_service_tagging) `service`, `env` and `version` as tagging best practices. Should I set `service:sidekiq` by default? Or should those three tags remain available for the app developer? What about client-side metrics which are triggered in Puma by Sidekiq client APIs? Should they use `service:sidekiq` or a `sidekiq.` prefix?

**Comment by weppos** [comment]
> Datadog does [document](https://docs.datadoghq.com/getting_started/tagging/unified_service_tagging) `service`, `env` and `version` as tagging best practices. 

I am not sure how this applies to metrics. I can see service being populated for log entries, but all the metrics I spot checked in our account do not have the `service`, nor `version`. As for `env` they do, but just because we actually configure each agent running on our systems to include a certain set of tags such as `env`, `datacenter`, and other. Here's an example of a Redis metric.

<img width="1369" alt="Image" src="https://github.com/user-attachments/assets/b420e179-adeb-401e-96ce-8d7b6fb44497" />

Actually, as you can see, it's standard for every software to have its own prefix. See some examples from Redis, Postgresql, and  Aws.

<img width="551" alt="Image" src="https://github.com/user-attachments/assets/87c9af14-3d75-46d7-ab84-9bd1d0c2f85e" />
<img width="506" alt="Image" src="https://github.com/user-attachments/assets/afb1bb95-da32-4319-8509-256bc4a8a295" />

Also note that, at least for what I can see in the logs dashboard, the service is not the software, but the actual instance (process). For example, I can have a software log originating from the software Nginx, and if I run 2 instances (processes), they will both be reported as `nginx.METRIC` whereas the `service` value (attached as tag/metadata) will be the actual service name such as `application1` and `application2`.

Finally, I'm bringing DD examples here, but also note that DD has quite advanced capabilities and it supports tagging. Not all metric consumers may be supporting tagging or such advanced features.

TL;DR: IMHO it is reasonable to prefix your metrics with `sidekiq.` to identify the source software. I also don't think you should set a service tag, because (1) not all consumers may support it, (2) tags increase metrics cardinality and increase costs (3) consumers may want to set their on service to match the running process and specifically tag which instance of sidekiq is running.

**Comment by mperham** [comment]
😎👍🏻

**Comment by mperham** [comment]
Thanks for all of the feedback, I truly appreciate it. All metrics will be prefixed with `sidekiq.` in 8.0.
